### DISCUSSION TITLE
Top level imports and type hints

### DISCUSSION THREAD

**Discussion by ampx-mg** [apache/airflow#36587] [Q&A] [answered]
Recently we have run into an issue, when trying to follow the top level imports best practice.

Say we have a function which I would like to reuse in different tasks. 
```
import pandas as pd

def pd_foo(df: pd.DataFrame):
  if df:
    print("OK")
```
Currently we have such function defined at the top level and therefore the import is at the top level as well, which breaks the best practice. Is there a better way to deal with this?
I was thinking I could copy the whole function inside each task that uses it, which for now seems fine, as only few tasks use it, but I wouldn't say this scales well. We could also drop the type hints for such cases, which I would rather avoid.  
Any tips are appreciated.

Furthermore we use KubernetesExecutor and for packages which are not included in the base install this becomes unfeasible, since our airflow server won't contain these packages, they are included in specific docker images which the jobs use. This isn't issue for `pandas`, but imagine using `tensorflow`. I would have to install the package on the server.

**Comment by potiuk** [comment, marked as answer]
Use `if TYPE_CHECKING` and either specify your type in quotes `'pd.Dataframe'` or use `from __future__ import annotations`.

Look those up - this is standard way you handle these problems in Python, nothing Airflow specific.

**Reply by ampx-mg** [reply]
Brilliant, thank you!

**Comment by potiuk** [comment]
> Furthermore we use KubernetesExecutor and for packages which are not included in the base install this becomes unfeasible, since our airflow server won't contain these packages, they are included in specific docker images which the jobs use. This isn't issue for pandas, but imagine using tensorflow. I would have to install the package on the server.

that's no problem if you use the above - simply make sure both Airflow image and worker image have those packages installed. You can (should) build your custom image if needed.
